core/src/main/java/org/infinispan/cache/impl/CacheImpl.java

```diff
@@ -1411,32 +1390,29 @@ public final void putAll(Map<? extends K, ? extends V> map, long lifespan, TimeU
       putAll(map, metadata);
    }
 
-   final void putAll(Map<? extends K, ? extends V> map, Metadata metadata, long explicitFlags, InvocationContext ctx) {
+   final void putAll(Map<? extends K, ? extends V> map, Metadata metadata, long explicitFlags, ContextBuilder contextBuilder) {
       // Vanilla PutMapCommand returns previous values; add IGNORE_RETURN_VALUES as the API will drop the return value.
       // Interceptors are free to clear this flag if appropriate (since interceptors are the only consumers of the
       // return value).
       explicitFlags = EnumUtil.mergeBitSets(explicitFlags, FlagBitSets.IGNORE_RETURN_VALUES);
-      PutMapCommand command = createPutAllCommand(map, metadata, explicitFlags, ctx);
-      executeCommandAndCommitIfNeeded(ctx, command);
+      PutMapCommand command = createPutAllCommand(map, metadata, explicitFlags);
+      executeCommandAndCommitIfNeeded(contextBuilder, command);
    }
 
    public final Map<K, V> getAndPutAll(Map<? extends K, ? extends V> map) {
-      return getAndPutAll(map, defaultMetadata, EnumUtil.EMPTY_BIT_SET, getInvocationContextWithImplicitTransaction(false, map.size()));
+      return getAndPutAll(map, defaultMetadata, EnumUtil.EMPTY_BIT_SET, () -> getInvocationContextWithImplicitTransaction(map.size()));
```

Comment:
I wonder if sizing the lookupEntries in context justifies this allocation enough - I would rather drop it and create multi-key context all the time (and if you want to optimize the single-key case you could supply the proper method handle right here - though I think it's an unnecessary optimization).

Reply:
I did some refactorization. now the context builder receives the key count as parameter